libraries/AP_Networking/AP_Networking_PPP.cpp

```diff
@@ -0,0 +1,197 @@
+
+#include "AP_Networking_Config.h"
+
+#if AP_NETWORKING_BACKEND_PPP
+
+#include "AP_Networking_PPP.h"
+#include <GCS_MAVLink/GCS.h>
+
+#include <lwip/udp.h>
+#include <lwip/ip_addr.h>
+#include <netif/ppp/ppp_opts.h>
+#include <netif/ppp/pppapi.h>
+#include <netif/ppp/pppos.h>
+#include <lwip/tcpip.h>
+#include <stdio.h>
+
+
+extern const AP_HAL::HAL& hal;
+
+#if LWIP_TCPIP_CORE_LOCKING
+#define LWIP_TCPIP_LOCK() sys_lock_tcpip_core()
+#define LWIP_TCPIP_UNLOCK() sys_unlock_tcpip_core()
+#else
+#define LWIP_TCPIP_LOCK()
+#define LWIP_TCPIP_UNLOCK()
+#endif
+
+/*
+  output some data to the uart
+ */
+uint32_t AP_Networking_PPP::ppp_output_cb(ppp_pcb *pcb, const void *data, uint32_t len, void *ctx)
+{
+    auto &driver = *(AP_Networking_PPP *)ctx;
+    LWIP_UNUSED_ARG(pcb);
+    uint32_t remaining = len;
+    const uint8_t *ptr = (const uint8_t *)data;
+    while (remaining > 0) {
+        auto n = driver.uart->write(ptr, remaining);
+        if (n > 0) {
+            remaining -= n;
+            ptr += n;
+        } else {
+            hal.scheduler->delay_microseconds(100);
+        }
+    }
+    return len;
```

Comment:
There should be some sort of timeout on this because if you disconnect while writing multiple packets it could delay forever... In which it would then return `remaining - len`.

Reply:
i did originally return short, but PPP in lwip then throws an assertion, so all we can do is keep trying